### PR TITLE
Allow user to undo single step

### PR DESCRIPTION
**Pull request by tuankiet65**
This pull request adds a button to allow user to undo a single step.

The button is initially disabled after a new game. After a move is made the button is enabled, and after user undoes a move, the button is disabled again. User can undo even after encountered a Game Over.

This pull request implements #9.

### CODE REVIEW

File: Game.cpp [hunk 1]
```diff
@@ -8,6 +8,7 @@
 
 #include <algorithm>
 #include <cstdlib>
+#include <cstring>
```
Comment (humdingerb) [inline Game.cpp]: Really needed?

File: WindowBoard.cpp [hunk 2]
```diff
@@ -160,21 +186,28 @@ WindowBoard::~WindowBoard()
...
-	messenger.SendMessage(H2048_WINDOW_SHOW);
+	messenger.SendMessage(&redraw);
+
 	fSending = true;
 }
 
 void
 WindowBoard::gameEnded()
-{
+{
```
Comment (humdingerb) [inline WindowBoard.cpp]: Stray tab

**Comment by scottmc** [comment]
Instead of adding an undo button, I would add it as an option on a menu of the menubar.  If you put an undo button in front of the user, they are more likely to use it more often to cheat.

**Comment by humdingerb** [comment]
> Instead of adding an undo button, I would add it as an option on a menu of the menubar.

How about making that another GCI task?
If we go for that, we should clean up the look in general. Put the score in the window title and move the "New game" and "Undo last move" into the menu (together with a "Quit" and "About" item).

File: Game.cpp (hunk 1 above)
Comment (tuankiet65) [reply Game.cpp]: Initially I intended to use std::memcpy to copy the tables, but then I switched to a for-loop and I forgot to remove that header. My mistakes.

**Comment by tuankiet65** [comment]
> (Not sure how the commandline version is supposed to work.)
I've added a 'u' key in the command-line version, not the GUI though.

**Comment by humdingerb** [comment]
> I've added a 'u' key for undo in the command-line version, not the GUI though. I'll work on it tomorrow since it's already midnight in my place.

Can you add the shortcut to the GUI, too?
As you play the game with the keyboard, having a shortcut is vital, or you have to hassle with grabbing the mouse and point and click on the button.

**Comment by tuankiet65** [comment]
I've added it in 2042969.

**Comment by humdingerb** [comment]
Works, thanks!
I'll wait with the merge if some better-than-me-coder has some input.

**Comment by humdingerb** [comment]
Undoing a step doesn't decrease the score accordingly. Sorry I didn't notice sooner...
Can you try to fix that, @tuankiet65 ?

**Comment by tuankiet65** [comment]
Hmm that's weird, it works on my side (at least on my branch). I can't test master because it doesn't compile (#18, last comment)

**Comment by humdingerb** [comment]
You right. It does work... Probably didn't relize those were "non-scoring" moves that - of course - don't change the score when reverted. Sorry... :)

**Comment by tuankiet65** [comment]
No problem :)

**Pull request merged**